### DISCUSSION TITLE
aDetailer not functioning in 16bit precision via DirectML

### DISCUSSION THREAD

**Discussion by brknsoul** [Bing-su/adetailer#188] [General]
I have an AMD RX 6600 XT 8GB GPU in Windows 10, thus I need to use DirectML.

If I run SD in 16bit precision aDetailer doesn't seem to function; it does detect faces and attempt to fix them (as evidenced by console output), but the faces don't change.

Once switching to 32bit precision, aDetailer inpaints faces correctly.

Although this seems to be a bug, I thought I'd raise it as a General discussion first.

Now I just have to make some samples.

All aDetailer functions performed using "mediapipe_full" detection model, with default (unchanged) settings.

32bit precision, control (aDetailer OFF)
![00147-20230705213400-701237304](https://github.com/Bing-su/adetailer/assets/1969381/4296f6b9-e81c-48e0-a848-6503b024b90a)![image](https://github.com/Bing-su/adetailer/assets/1969381/28f2e692-23eb-4598-a2d8-812896e5d311)

32bit precision, aDetailer ON
![00146-20230705213326-701237304](https://github.com/Bing-su/adetailer/assets/1969381/4d7d2c3b-123c-424b-9ae5-aed8f55fa229)

16bit precision, control (aDetailer OFF)
![00148-20230705214122-701237304](https://github.com/Bing-su/adetailer/assets/1969381/822f7f86-7e11-46fa-b7df-6acea25e567f)![image](https://github.com/Bing-su/adetailer/assets/1969381/50634fe7-3338-4511-8e1f-d0cc50bdbd84)

16bit precision, aDetailer ON
![00149-20230705214224-701237304](https://github.com/Bing-su/adetailer/assets/1969381/3c0eca48-2203-42df-9bea-0a3afeac4815)

**Comment by Bing-su** [comment]
If you see `NansException` in your console log, this is the intended behavior.

**Comment by brknsoul** [comment]
I'm getting no `NansException` in console.

```
Loading weights: E:\automatic\models\Stable-diffusion\aZovyaPhotoreal_v2.safetensors --------------- 0.0/2.1 GB -:--:--
08:55:17-906148 INFO     Setting Torch parameters: dtype=torch.float16 vae=torch.float16 unet=torch.float16
LatentDiffusion: Running in eps-prediction mode
DiffusionWrapper has 859.52 M params.
08:55:49-663369 INFO     Loading VAE weights: specified in settings: cached vae-ft-mse-840000-ema-pruned.safetensors
08:55:52-663563 INFO     Applying split attention optimization
08:55:52-833554 INFO     Model loaded in 35.4s (load=0.4s create=0.7s apply=25.9s vae=6.9s move=1.2s embeddings=0.2s)
08:55:54-398711 INFO     Model load finished: {'ram': {'used': 4.82, 'total': 15.95}} cached=2
100%|██████████████████████████████████████████████████████████████████████████████████| 25/25 [00:19<00:00,  1.26it/s]
mediapipe: 1 detected.
100%|██████████████████████████████████████████████████████████████████████████████████| 11/11 [00:08<00:00,  1.23it/s]
```

![image](https://github.com/Bing-su/adetailer/assets/1969381/81680b7e-2479-47a3-b13a-c2a15c6c9ba1)

**Comment by Bing-su** [comment]
Then turn on the mask preview option, and compare the preview to the generated image.

**Comment by brknsoul** [comment]
16 bit
![00105-20230707132710-701237304-ad-preview](https://github.com/Bing-su/adetailer/assets/1969381/849bfea9-adf8-4146-9e39-653f05f5bb43)![00107-20230707132723-701237304](https://github.com/Bing-su/adetailer/assets/1969381/f17daeed-f17b-4923-ae99-8b341c339a24)

32bit
![00108-20230707133254-701237304-ad-preview](https://github.com/Bing-su/adetailer/assets/1969381/4bafbdf0-a044-451a-9950-54ac97e7695f)![00110-20230707133309-701237304](https://github.com/Bing-su/adetailer/assets/1969381/a6e1ef2c-44af-4a0e-9d31-70d6ce49baf8)

**Comment by Bing-su** [comment]
Sorry, but could you turn on the save image before adetailer option and compare it to this image as well?

**Comment by brknsoul** [comment]
I did do that;
16bit;
![00106-20230707132723-701237304-ad-before](https://github.com/Bing-su/adetailer/assets/1969381/9c816929-e53a-428f-85aa-233945d5239e)

32bit
![00109-20230707133308-701237304-ad-before](https://github.com/Bing-su/adetailer/assets/1969381/144bea11-c44f-49c1-b9b5-cb8b6164e8c7)

**Comment by Bing-su** [comment]
No, I didn't mean to do an on-off comparison, I meant to post images with `-ad-before` in the filename.

**Reply by brknsoul** [reply]
Those are. Each generation produces 3 images; masked, before, and after.

![image](https://github.com/Bing-su/adetailer/assets/1969381/1eb3058f-af9d-4117-90b7-dfc20021ef38)

**Comment by Bing-su** [comment]
Maybe it's a directML problem.
https://github.com/lshqqytiger/stable-diffusion-webui-directml/issues/87
https://github.com/vladmandic/automatic/issues/984

**Reply by brknsoul** [reply]
Alright. I thought it might be. Just making sure it's not something on your end.

Thanks for your help.. and the awesome extension! :-)
